src/main/cpp/js.cpp

```diff
@@ -0,0 +1,1071 @@
+#include <vector>
+
+#include "ulight/impl/buffer.hpp"
+#include "ulight/impl/highlight.hpp"
+#include "ulight/ulight.hpp"
+
+#include "ulight/impl/chars.hpp"
+#include "ulight/impl/js.hpp"
+#include "ulight/impl/unicode.hpp"
+
+namespace ulight {
+
+namespace js {
+
+#define ULIGHT_JS_TOKEN_TYPE_U8_CODE(id, code, highlight, source) u8##code,
+#define ULIGHT_JS_TOKEN_TYPE_LENGTH(id, code, highlight, source) (sizeof(u8##code) - 1),
+#define ULIGHT_JS_TOKEN_HIGHLIGHT_TYPE(id, code, highlight, source) (Highlight_Type::highlight),
+#define ULIGHT_JS_TOKEN_TYPE_FEATURE_SOURCE(id, code, highlight, source) (Feature_Source::source),
+
+namespace {
+inline constexpr std::u8string_view token_type_codes[] {
+    ULIGHT_JS_TOKEN_ENUM_DATA(ULIGHT_JS_TOKEN_TYPE_U8_CODE)
+};
+
+static_assert(std::ranges::is_sorted(token_type_codes));
+
+inline constexpr unsigned char token_type_lengths[] {
+    ULIGHT_JS_TOKEN_ENUM_DATA(ULIGHT_JS_TOKEN_TYPE_LENGTH)
+};
+
+inline constexpr Highlight_Type token_type_highlights[] {
+    ULIGHT_JS_TOKEN_ENUM_DATA(ULIGHT_JS_TOKEN_HIGHLIGHT_TYPE)
+};
+
+inline constexpr Feature_Source token_type_sources[] {
+    ULIGHT_JS_TOKEN_ENUM_DATA(ULIGHT_JS_TOKEN_TYPE_FEATURE_SOURCE)
+};
+} // namespace
+
+/// @brief Returns the in-code representation of `type`.
+[[nodiscard]]
+std::u8string_view js_token_type_code(Token_Type type) noexcept
+{
+    return token_type_codes[std::size_t(type)];
+}
+
+/// @brief Equivalent to `js_token_type_code(type).length()`.
+[[nodiscard]]
+std::size_t js_token_type_length(Token_Type type) noexcept
+{
+    return token_type_lengths[std::size_t(type)];
+}
+
+[[nodiscard]]
+Highlight_Type js_token_type_highlight(Token_Type type) noexcept
+{
+    return token_type_highlights[std::size_t(type)];
+}
+
+[[nodiscard]]
+Feature_Source js_token_type_source(Token_Type type) noexcept
+{
+    return token_type_sources[std::size_t(type)];
+}
+
+[[nodiscard]]
+std::optional<Token_Type> js_token_type_by_code(std::u8string_view code) noexcept
+{
+    const std::u8string_view* const result = std::ranges::lower_bound(token_type_codes, code);
+    if (result == std::end(token_type_codes) || *result != code) {
+        return {};
+    }
+    return Token_Type(result - token_type_codes);
+}
+
+std::size_t match_whitespace(std::u8string_view str)
+{
+    constexpr auto predicate = [](char8_t c) { return is_js_whitespace(c); };
+    return std::size_t(std::ranges::find_if_not(str, predicate) - str.begin());
+}
+
+std::size_t match_non_whitespace(std::u8string_view str)
+{
+    constexpr auto predicate = [](char8_t c) { return !is_js_whitespace(c); };
+    return std::size_t(std::ranges::find_if_not(str, predicate) - str.begin());
+}
+
+std::size_t match_line_comment(std::u8string_view s) noexcept
+{
+    if (!s.starts_with(u8"//")) {
+        return 0;
+    }
+
+    // Skip the '//' prefix
+    std::size_t length = 2;
+
+    // Continue until EoL.
+    while (length < s.length()) {
+        if (s[length] == u8'\n') {
```

Comment:
https://262.ecma-international.org/15.0/index.html#prod-LineTerminator

This is not quite correct; there are more characters in JS that can terminate a line. To avoid doing a full UTF-8 decode here, you can add a function as follows:
```cpp
std::size_t match_line_terminator(std::u8string_view s)
{
    constexpr std::u8string_view crlf = u8"\r\n";
    constexpr std::u8string_view ls = u8"\u2028";
    constexpr std::u8string_view ps = u8"\u2029";

    return s.starts_with(u8'\n') ? 1
        : s.starts_with(crlf) ? crlf.length()
        : s.starts_with(ls) ? ls.length()
        : s.starts_with(ps) ? ps.length()
        : 0;
}
```
Then:
```suggestion
        if (match_line_separator(s.substr(length))) {
```
I'm not actually sure if the length value returned from `match_line_separator` is needed somewhere, but may as well have it.